skfda/preprocessing/dim_reduction/_pace.py

```diff
@@ -0,0 +1,1441 @@
+                pairs.
+            r_eval: Array of shape (n_eval, d) with r-direction evaluation
+                points.
+            s_eval: Array of shape (n_eval, d) with s-direction evaluation
+                points.
+            n_eval: Number of evaluation points
+                (i.e., len(r_eval) == len(s_eval)).
+            n_obs: Number of observed point pairs.
+
+        Returns:
+            Array of shape (n_eval, n_eval, n_obs, 3) representing the design
+            matrix with columns [1, t_r - r_eval, t_s - s_eval].
+        """
+        x = np.ones((n_eval, n_eval, n_obs, 3))
+        for i in range(n_eval):
+            for j in range(n_eval):
+                x[:, j, :, 1, None] = t_pairs[None, :, 0] - r_eval[:, None]
+                x[i, :, :, 2, None] = t_pairs[None, :, 1] - s_eval[:, None]
+        return x
+
+    def _cov_lls(
+        self,
+        h: float,
+        r_eval: NDArrayFloat,
+        s_eval: NDArrayFloat,
+        cov_coords: NDArrayFloat,
+        cov_values: NDArrayFloat,
+        win: NDArrayFloat,
+    ) -> NDArrayFloat:
+        """
+        Local linear smoother for covariance estimation.
+
+        Args:
+            h: Bandwidth for the kernel.
+            r_eval: First array of query points where smoother is evaluated.
+            s_eval: Second array of query points where smoother is evaluated.
+            cov_coords: Coordinates of the covariance.
+            cov_values: Values of the covariance.
+            win: Weights for the covariance.
+
+        Returns:
+            n_grid_points x n_grid_points array of smoothed covariance values.
+        """
+        active = np.nonzero(win)[0]
+        t_pairs = cov_coords[active, :]
+        cov_values = cov_values[active]
+        win = win[active]
+
+        n_eval, _ = r_eval.shape
+        n_obs, _ = cov_values.shape
+
+        weights = self._compute_cov_weights(h, r_eval, s_eval, t_pairs, win)
+        x = self._build_design_matrix(t_pairs, r_eval, s_eval, n_eval, n_obs)
+
+        x_t = np.transpose(x, (0, 1, 3, 2))
+        xtw = x_t * weights[:, :, None, :]
+        xtwx = xtw @ x
+        xtwy = xtw @ cov_values
+
+        beta = np.linalg.pinv(xtwx) @ xtwy
+
+        cov = beta[:, :, 0]
+        cov_t = np.transpose(cov, (1, 0, 2))
+        return np.array((cov + cov_t) / 2.0)  # noqa: WPS432
+
+    def _sort_and_clip_eigenpairs(
+        self,
+        eigenvalues: NDArrayFloat,
+        eigenvectors: NDArrayFloat,
+    ) -> tuple[NDArrayFloat, NDArrayFloat]:
+        """Sort and non-negatively clip eigenvalues, reorder eigenvectors."""
+        eigenvalues = np.maximum(eigenvalues, 0)
+        idx = np.argsort(eigenvalues)[::-1]
+        return eigenvalues[idx], eigenvectors[:, idx]
+
+    def _normalize_and_orient(
+        self,
+        eigenvectors: NDArrayFloat,
+        t: NDArrayFloat,
+    ) -> NDArrayFloat:
+        """Normalize and align the eigenvectors."""
+        for i in range(eigenvectors.shape[1]):
```

Comment:
I think the eigenvectors are already normalized. If normalization were needed, we could do this in a vectorized way, not in a Python loop. Also, if you want to adjust the sign for determinism, maybe it is better to use `svd_flip`, like we do for functional diffusion maps.